File: day12/day12.py
#!/usr/bin/env python3

import re
import math

class Moon(object):
    def __init__(self, pos):
        self.pos = pos
        self.vel = [0, 0, 0]

pattern = re.compile(r"<x=(-?[0-9]+), y=(-?[0-9]+), z=(-?[0-9]+)>")
def load_moons(text):
    moons = []
    for line in text.split("\n"):
        m = pattern.search(line)
        if m:
            moons.append(Moon([int(x) for x in m.groups()]))
    return moons

def update_step(pos, vel, length):
    # update the velocity
    for i in range(length):
        for j in range(i+1, length):
            if pos[i] < pos[j]:
                vel[i] += 1
                vel[j] -= 1
            elif pos[i] > pos[j]:
                vel[i] -= 1
                vel[j] += 1
            else:
                pass

    # update the position
    for i in range(length):
        pos[i] += vel[i]

def energy(moons, steps):
    for axis in range(3):
        pos = [m.pos[axis] for m in moons]
        vel = [m.vel[axis] for m in moons]
        for i in range(steps):
            update_step(pos, vel, len(moons))

        for i in range(len(moons)):
            moons[i].pos[axis] = pos[i]
            moons[i].vel[axis] = vel[i]

    energy = 0
    for m in moons:
        p = sum(abs(i) for i in m.pos)
        k = sum(abs(i) for i in m.vel)
        energy += k*p

    return energy

def repeat_axis(moons, axis):
    pos = [m.pos[axis] for m in moons]
    vel = [m.vel[axis] for m in moons]
    orig_pos = tuple(pos)
    orig_vel = tuple(vel)

    length = len(moons)
    count = 0
    while True:
        update_step(pos, vel, length)
        count += 1

        if tuple(pos) == orig_pos and tuple(vel) == orig_vel:
            break

    return count

def lcm(a, b):
    return a * b // math.gcd(a, b)

def repeat(moons):
    x = repeat_axis(moons, 0)
    y = repeat_axis(moons, 1)
    z = repeat_axis(moons, 2)
    return lcm(lcm(x, y), z)

moons = load_moons("""
<x=-1, y=0, z=2>
<x=2, y=-10, z=-7>
<x=4, y=-8, z=8>
<x=3, y=5, z=-1>
""")
assert energy(moons, 10) == 179
assert repeat(moons) == 2772

moons = load_moons("""
<x=-8, y=-10, z=0>
<x=5, y=5, z=10>
<x=2, y=-7, z=3>
<x=9, y=-8, z=-3>
""")
assert energy(moons, 100) == 1940
assert repeat(moons) == 4686774924

with open("input", "rt") as f:
    moons = load_moons(f.read())

print("part1:", energy(moons, 1000))
print("part2:", repeat(moons))

File: day12/day12.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <regex.h>

struct moon
{
	int64_t pos[3];
	int64_t vel[3];
};

struct moon *load_moons(FILE *input, size_t *count)
{
	regex_t regex;
	const char *pattern = "<x=(-?[0-9]+), y=(-?[0-9]+), z=(-?[0-9]+)>";
	if (regcomp(&regex, pattern, REG_EXTENDED))
	{
		return NULL;
	}
	struct moon *m = NULL;
	size_t msize = 0;
	size_t mcount = 0;

	char *line = NULL;
	size_t lsize = 0;
	while (getline(&line, &lsize, input) != -1)
	{
		regmatch_t match[4];
		if (regexec(&regex, line, 4, match, 0))
		{
			continue;
		}
		if (mcount == msize)
		{
			size_t nsize = msize ? msize * 2 : 4;
			struct moon *nm = realloc(m, nsize * sizeof(*m));
			if (!nm)
			{
				regfree(&regex);
				free(m);
				return NULL;
			}
			msize = nsize;
			m = nm;
		}
		for (int i = 0; i < 3; i++)
		{
			m[mcount].pos[i] = strtol(line+match[i+1].rm_so, NULL, 10);
			m[mcount].vel[i] = 0;
		}
		mcount++;
	}
	free(line);
	regfree(&regex);

	*count = mcount;
	return m;
}

void step_axis(struct moon *m, size_t count, int axis)
{
	for (size_t i = 0; i < count; i++)
	{
		for (size_t j = i + 1; j < count; j++)
		{
			if (m[i].pos[axis] < m[j].pos[axis])
			{
				m[i].vel[axis]++;
				m[j].vel[axis]--;
			}
			else if (m[i].pos[axis] > m[j].pos[axis])
			{
				m[i].vel[axis]--;
				m[j].vel[axis]++;
			}
		}
	}
	for (size_t i = 0; i < count; i++)
	{
		m[i].pos[axis] += m[i].vel[axis];
	}
}

int64_t energy(struct moon *m, size_t count, int steps)
{
	for (int a = 0; a < 3; a++)
	{
		for (int j = 0; j < steps; j++)
		{
			step_axis(m, count, a);
		}
	}
	int64_t energy = 0;
	for (size_t i = 0; i < count; i++)
	{
		int64_t pot = 0;
		int64_t kin = 0;
		for (int a = 0; a < 3; a++)
		{
			pot += m[i].pos[a] > 0 ? m[i].pos[a] : -m[i].pos[a];
			kin += m[i].vel[a] > 0 ? m[i].vel[a] : -m[i].vel[a];
		}
		energy += pot * kin;
	}
	return energy;
}

int64_t repeat_axis(struct moon *m, size_t count, int axis)
{
	int64_t *pos = malloc(2 * count *sizeof(*pos));
	int64_t *vel = pos + count;
	for (size_t i = 0; i < count; i++)
	{
		pos[i] = m[i].pos[axis];
		vel[i] = m[i].vel[axis];
	}

	int64_t rep = 0;
	for (;;)
	{
		step_axis(m, count, axis);
		rep++;
		size_t i;
		for (i = 0; i < count; i++)
		{
			if (m[i].pos[axis] != pos[i] || m[i].vel[axis] != vel[i])
			{
				break;
			}
		}
		if (i == count)
		{
			break;
		}
	}
	free(pos);
	return rep;
}

int64_t gcd(int64_t a, int64_t b)
{
	while (b)
	{
		int t = b;
		b = a % b;
		a = t;
	}
	return a;
}

int64_t lcm(int64_t a, int64_t b)
{
	return a*b / gcd(a, b);
}

int64_t repeat(struct moon *m, size_t count)
{
	int64_t x = repeat_axis(m, count, 0);
	int64_t y = repeat_axis(m, count, 1);
	int64_t z = repeat_axis(m, count, 2);
	return lcm(lcm(x, y), z);
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <input>\n", argv[0]);
		return -1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "Cannot open %s for reading\n", argv[1]);
		return -1;
	}

	size_t count;
	struct moon *moons = load_moons(input, &count);
	fclose(input);
	printf("part1: %ld\n", energy(moons, count, 1000));
	printf("part2: %ld\n", repeat(moons, count));
	free(moons);
	return 0;
}
